### PR TITLE
DOMA-1626 meter schema improvements

### PR DESCRIPTION
**Pull request by nomerdvadcatpyat**
- **Fix access rights:**
**Problem:** a user with `resident` type can access all `meters` and `meterReadings` of his `resinents` if he has at least 1 `resident` with `serviceConsumer` whose `accountNumber` matches `meter.accountNumber`.
**Fix:**  Now a user with `resident` type needs to pass `property` and `unitName` in the request for `meters`, and `property`, `unitName` and `meter` in the request for `metersReading`. Based on the request, it will be possible to find the right resident in the way and give only the necessary data.If a user with the `resident` type does not pass the `unitName` and `property` to `meters` query, then an empty array is returned to him.

---

- Add virtual field `hasMeters` in `resident.organizationFeatures`
**Problem:**: a mobile application needs to understand when an `organization` has `meters` (even before creating a `serviceConsumer`)
**Fix:** add virtual field `hasMeters` in `resident.organizationFeatures`

### CODE REVIEW

File: apps/condo/domains/meter/access/Meter.js [hunk 1]
```diff
@@ -19,24 +20,22 @@ async function canReadMeters ({ authentication: { item: user }, context }) {
...
-            if (meters.length > 0) return {}
+        const { property, unitName } = getMetersDataFromRequestBody(context)
+        const [resident] = await ResidentServerUtils.getAll(context, {
+            user: { id: userId },
+            property,
+            unitName,
+        })
+        const residentId = get(resident, 'id', null)
+        const serviceConsumers = await ServiceConsumer.getAll(context, {
+            resident: { id: residentId },
```
Comment (pahaz) [inline apps/condo/domains/meter/access/Meter.js:31]: deletedAd: null

File: apps/condo/domains/meter/access/Meter.js [hunk 2]
```diff
@@ -19,24 +20,22 @@ async function canReadMeters ({ authentication: { item: user }, context }) {
...
-
-            const meters = await Meter.getAll(context, {
-                property: { id: residentPropertyId },
-                unitName: residentUnitName,
-                accountNumber_in: serviceConsumerAccounts,
-            })
-
-            if (meters.length > 0) return {}
+        const { property, unitName } = getMetersDataFromRequestBody(context)
+        const [resident] = await ResidentServerUtils.getAll(context, {
```
Comment (pahaz) [inline apps/condo/domains/meter/access/Meter.js:24]: deletedAt: null

File: apps/condo/domains/meter/access/MeterReading.js [hunk 3]
```diff
@@ -10,31 +12,31 @@ const { queryOrganizationEmployeeFromRelatedOrganizationFor, queryOrganizationEm
...
-                resident: { id: resident.id },
-            })
-            const serviceConsumerAccounts = serviceConsumers.map(serviceConsumer => serviceConsumer.accountNumber)
-
-            const meters = await Meter.getAll(context, {
-                property: { id: residentPropertyId },
-                unitName: residentUnitName,
+        const { property, unitName, meter } = getMetersDataFromRequestBody(context)
+        const [resident] = await ResidentServerUtils.getAll(context, {
+            user: { id: userId },
```
Comment (pahaz) [inline apps/condo/domains/meter/access/MeterReading.js:23]: deletedAt: null

File: apps/condo/domains/meter/access/MeterReading.js [hunk 4]
```diff
@@ -10,31 +12,31 @@ const { queryOrganizationEmployeeFromRelatedOrganizationFor, queryOrganizationEm
...
+            property,
+            unitName,
+        })
+        const residentId = get(resident, 'id', null)
+        const serviceConsumers = await ServiceConsumer.getAll(context, {
+            resident: { id: residentId },
```
Comment (pahaz) [inline apps/condo/domains/meter/access/MeterReading.js:29]: deletedAt

File: apps/condo/domains/meter/access/MeterReading.js [hunk 5]
```diff
@@ -57,40 +59,41 @@ async function canManageMeterReadings ({ authentication: { item: user }, origina
...
+            })
+            const serviceConsumerAccounts = serviceConsumers.map(serviceConsumer => serviceConsumer.accountNumber)
 
-                const meters = await Meter.getAll(context, {
-                    property: { id: residentPropertyId },
-                    unitName: residentUnitName,
-                    accountNumber_in: serviceConsumerAccounts,
-                })
+            const meters = await Meter.getAll(context, {
+                id: meterId,
```
Comment (pahaz) [inline apps/condo/domains/meter/access/MeterReading.js:90]: deletedAt

File: apps/condo/domains/meter/access/MeterReading.js [hunk 6]
```diff
@@ -57,40 +59,41 @@ async function canManageMeterReadings ({ authentication: { item: user }, origina
...
+            const residentPropertyId = get(resident, ['property', 'id'], null)
+            const residentUnitName = get(resident, 'unitName', null)
+            const residentId = get(resident, 'id', null)
 
-                const serviceConsumers = await ServiceConsumer.getAll(context, {
-                    resident: { id: resident.id },
-                })
-                const serviceConsumerAccounts = serviceConsumers.map(serviceConsumer => serviceConsumer.accountNumber)
+            const serviceConsumers = await ServiceConsumer.getAll(context, {
+                resident: { id: residentId },
```
Comment (pahaz) [inline apps/condo/domains/meter/access/MeterReading.js:85]: deletedAt

**Review by toplenboren — APPROVED** [review]

File: apps/condo/domains/resident/schema/Resident.test.js [hunk 7]
```diff
@@ -317,6 +322,85 @@ describe('Resident', () => {
...
+                const [resource] = await MeterResource.getAll(client, { id: COLD_WATER_METER_RESOURCE_ID })
+                await createTestMeter(adminClient, organization, property, resource, {
+                    unitName,
+                })
+
+                const [obj] = await Resident.getAll(client, { id: resident.id })
+                expect(obj.organizationFeatures.hasMeters).toBe(true)
+            })
+
+            it('correctly sets the hasMeters: "false" if meters no meters existing in resident property and unit', async () => {
```
Comment (AntonAL) [inline apps/condo/domains/resident/schema/Resident.test.js:345]:
```suggestion
            it('correctly sets the hasMeters: "false" if no meters existing in resident property and unit', async () => {
```

File: apps/condo/domains/meter/access/MeterReading.js [hunk 8]
```diff
@@ -10,31 +12,38 @@ const { queryOrganizationEmployeeFromRelatedOrganizationFor, queryOrganizationEm
...
+        const serviceConsumers = await ServiceConsumer.getAll(context, {
+            resident: { id: residentId },
+            deletedAt: null,
+        })
+        const serviceConsumerAccounts = serviceConsumers.map(serviceConsumer => serviceConsumer.accountNumber)
+
+        return {
+            meter: {
+                ...meter,
+                property: { id: propertyId },
```
Comment (AntonAL) [inline apps/condo/domains/meter/access/MeterReading.js:43]: What if property will be deleted?

**Review by AntonAL — APPROVED** [review]

File: apps/condo/domains/meter/access/MeterReading.js (hunk 8 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/meter/access/MeterReading.js:43]: now there is a MeterReading on the deleted property, added deletedAt

File: apps/condo/domains/meter/access/MeterReading.js [hunk 9]
```diff
@@ -10,31 +12,40 @@ const { queryOrganizationEmployeeFromRelatedOrganizationFor, queryOrganizationEm
...
-                resident: { id: resident.id },
-            })
-            const serviceConsumerAccounts = serviceConsumers.map(serviceConsumer => serviceConsumer.accountNumber)
-
-            const meters = await Meter.getAll(context, {
-                property: { id: residentPropertyId },
-                unitName: residentUnitName,
+        const { meter: meterQuery } = getMetersDataFromRequestBody(context)
+        const [meter] = await Meter.getAll(context, {
+            ...meterQuery,
```
Comment (pahaz) [inline apps/condo/domains/meter/access/MeterReading.js:23]: deletedAt: null

**Pull request merged**